Some improvements for braids computations

enriqueartal opened the pull request:
### 📚 Description
Fasten the computation of the left normal form of a braid, produce a shorter conjugating braid and if possible a pure conjugating braid

- The method `left_normal_form` for braids was modified to be part of the general method for Artin groups. I propose to come back to the former method, it produces the same result but much faster. I added `left_normal_form` as a method for braid group and kept `_left_normal_form_coxeter` though it is not used any more for braids.
- The method `conjugating_braid`  produces a braid in left normal form with high powers of `Delta`, which can be reduce to 0 or 1.
- If two braids are conjugated and have the same permutation, the conjugation can be realized by a pure braid. Starting from `conjugating_braid` a simple computation produces such a pure braid.

There are mild improvements for the treatment of braid groups. 

Resolves #35213


### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### ⌛ Dependencies
<!-- List all open pull requests that this PR logically depends on -->
<!--
- #xyz: short description why this is a dependency
- #abc: ...
-->

enriqueartal committed:
conjugating

enriqueartal committed:
first version of a shorter conjugating_braid

enriqueartal committed:
adding pure_conjugating_braid method

enriqueartal committed:
some trail white spaces

enriqueartal committed:
add left_normal_form and simplify previous code

trevorkarn commented on src/sage/groups/braid.py:
```diff
@@ -1517,8 +1539,10 @@ def _left_normal_form_coxeter(self):
-        """
-        Return the right normal form of the braid.
+        r"""
+        A tuple of simple generators in the right normal form. The last
```
You could leave the original "Return the right normal form" and have your additions on the next line after a blank line.

trevorkarn commented on src/sage/groups/braid.py:
```diff
@@ -1620,21 +1644,31 @@ def conjugating_braid(self, other):
             sage: B = BraidGroup(3)
             sage: a = B([2, 2, -1, -1])
-            sage: b = B([2, 1, 2, 1])
+            sage: b = B([2, 2, 2, 2, 1])
```
Maybe I'm not understanding something, but you could leave the already existing tests and just add to them below.

trevorkarn commented on src/sage/groups/braid.py:
```diff
@@ -1447,6 +1448,26 @@ def annular_khovanov_homology(self, qagrad=None, ring=IntegerRing()):
+            sage: c = B([1])
+            sage: c.left_normal_form()
+            (1, s0)
+        """
```
You might consider adding tests for trivial examples.

enriqueartal replied:
Thanks for your comments. I completed the descriptions and added more examples.

enriqueartal committed:
Description and more examples

enriqueartal committed:
extra spaces removed

enriqueartal committed:
Move a comment from TODO to NOTE

enriqueartal committed:
More examples

enriqueartal committed:
- Correct error for conjugating_braid and pure_conjugating_braid when the left normal form was a power of Delta
  to ensure all the factors of a product are braids
- In pure_conjugating_braid, check if the permutation of the conjugating braid is in the subgroup generated by
  the permutations of the centralizing braids of self; return None if not.

enriqueartal committed:
Merge branch 'sagemath:develop' into braids

enriqueartal committed:
Merge branch 'sagemath:develop' into braids

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+        if p3 not in G.subgroup(LP):
+            return None
+        P = p3.word_problem(LP, display = False, as_list = True)
+        b1 = prod(LB[LP.index(G(a))] ** b for a,b in P)
```
What happens if `P` is empty? Add a doctest checking for this corner case (if it can occur).

tscrim replied:
Additionally, if you make `LP` a `dict` from permutations to braids, this becomes faster and you can still pass it off to `G.subgroup()`.

enriqueartal replied:
For the first question a previous `if` makes impossible to have `P` empty. I could not pass the dict to `subgroup` and even if it would be possible, I think I cannot pass it to `word_problem`

tscrim replied:
> For the first question a previous `if` makes impossible to have `P` empty.

I might say also the previous `p3.is_one()` also makes that the case. However, again this case should have a doctest.

> I could not pass the dict to `subgroup` and even if it would be possible, I think I cannot pass it to `word_problem`

You can pass a `dict` to `subgroup`:
```python
sage: G = groups.permutation.Symmetric(5)
sage: d = {G.random_element(): i for i in range(5)}
sage: d
{(1,5,2,4): 0, (1,5)(2,3): 1, (1,3,5,2): 2, (1,5,3,2): 3, (1,5)(2,4,3): 4}
sage: G.subgroup(d)
Subgroup generated by [(1,3,5,2), (1,5,3,2), (1,5)(2,3), (1,5)(2,4,3), (1,5,2,4)] of (Symmetric group of order 5! as a permutation group)
```

Indeed for `word_problem`, you cannot pass the `dict`, but you can do `list(d)`, which will have a list in the same order the `dict` `d` was constructed (Python3 guarantees this).

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+        L = leftnormalform(b0)
+        k = int(L[0][0] % 2)
+        b0 = B.delta() ** k * B(prod(B(a) for a in L[1:]))
+        return b0
```
```suggestion
        L = leftnormalform(b0)
        L[0][0] %= 2
        return B._element_from_libbraiding(L)
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+        LP = [G(a.permutation()) for a in LB]
+        if p3 not in G.subgroup(LP):
+            return None
+        P = p3.word_problem(LP, display = False, as_list = True)
```
PEP8
```suggestion
        P = p3.word_problem(LP, display=False, as_list=True)
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+    def pure_conjugating_braid(self, other):
+        r"""
+        Return a pure conjugating braid, if it exists.
+
```
Add a more detailed description; in particular, including the definition of a pure conjugating braid.

enriqueartal replied:
Done

tscrim replied:
It should not be part of the one-line description. Spend a little bit of time/mathematical notation spelling it out a bit more.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+
+        INPUT:
+
+        - ``other`` -- the other braid to look for conjugating braid
```
This is confusing.

enriqueartal replied:
Done

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1622,19 +1659,43 @@ def conjugating_braid(self, other):
+            n = B.strands()
+            k = int(l[0][0] % 2)
+            b0 = B.delta() ** k * B(prod(B(a) for a in l[1:]))
+            return b0
```
```suggestion
            l[0][0] %= 2
            return B._element_from_libbraiding(l)
```

enriqueartal replied:
Done

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1622,19 +1659,43 @@ def conjugating_braid(self, other):
+            sage: b = B([2, 2, 2, 2, 1])
+            sage: c = b * a / b
+            sage: d1 = a.conjugating_braid(c)
+            sage: print (len(d1.Tietze()))
```
PEP8
```suggestion
            sage: print(len(d1.Tietze()))
```
and similar below.

enriqueartal replied:
Done

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+            sage: print(a.pure_conjugating_braid(b))
+            None
+            sage: a1=B([1])
+            sage: a2=B([2])
```
PEP8
```suggestion
            sage: a1 = B([1])
            sage: a2 = B([2])
```

enriqueartal replied:
Done

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+        B = self.parent()
+        n = B.strands()
+        G = SymmetricGroup(n)
+        p1 = G(self.permutation())
```
I would only convert to `G` when it becomes necessary.

enriqueartal replied:
I move the definition of `G` few lines below, and eliminate some conversions

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1718,82 @@ def is_conjugated(self, other):
+        if p1 != p2:
+            return None
+        b0 = self.conjugating_braid(other)
+        if not b0:
```
```suggestion
        if b0 is not None:
```
It is better to check explicitly against `None`.

enriqueartal replied:
Done but putting `if b0 is None`

tscrim replied:
Ah, right, whoops. Thanks.

tscrim reviewed (changes requested):
You should not hide other computational methods but expose both through an, e.g., `algorithm` argument. This is not just good for testing purposes (which you should do a more systematic comparison between the two methods), but also gives additional options to the user.

enriqueartal committed:
Update src/sage/groups/braid.py

Simplify code for conjugating_braid

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Last leftnormalform in pure_conjugating_braid may produce long braids

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
more explanations

enriqueartal committed:
Check if the conjugating braid is shorter after left normal form

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1651,54 @@ def conjugating_braid(self, other):
         INPUT:
 
-        - ``other`` -- the other braid to look for conjugating braid
+        - ``other`` -- a braid in the same braid group as `self`.
```
```suggestion
        - ``other`` -- a braid in the same braid group as ``self``
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1651,54 @@ def conjugating_braid(self, other):
+
+        OUTPUT:
+
+        A braid `d` such that `other` equals `d^-1 * self * d`.
```
```suggestion
        A braid `d` such that `other` equals `d^{-1} x d`, where `x` is ``self``.
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1651,54 @@ def conjugating_braid(self, other):
-            True
-            sage: d
+            sage: d1 = a.conjugating_braid(c)
+            sage: print(d1)
```
You don't need the `print` here.
```suggestion
            sage: d1
```
Similar elsewhere.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1721,90 @@ def is_conjugated(self, other):
+
+        INPUT:
+
+        - ``other`` -- a braid in the same braid group as `self`.
```
```suggestion
        - ``other`` -- a braid in the same braid group as ``self``
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1721,90 @@ def is_conjugated(self, other):
+            return b2
+        else:
+            return b0
+
```
```suggestion
        return b2 if n2 <= n0 else b0
```

enriqueartal replied:
Done

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1721,90 @@ def is_conjugated(self, other):
+        B = self.parent()
+        n = B.strands()
+        p1 = self.permutation()
+        p2 = other.permutation()
```
Actually, since you are converting everything to elements of `SymmetricGroup`, it might be better just to either duplicate the `permutation` method (well, it redirects to `coxeter_group_element()`) or by expanding those methods to take an optional group argument. That way you have less transient objects. Although such changes are not needed for a positive review.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1651,54 @@ def conjugating_braid(self, other):
+            sage: print(len(d2.Tietze()))
+            13
+            sage: print(c.conjugating_braid(b))
+            None
```
```suggestion
            sage: c.conjugating_braid(b) is None
            True
```
(and similar) if you feel it is a better doctest. The current way is fine though but in my mind it is slightly more explicit.

enriqueartal replied:
Done

enriqueartal committed:
partial changes for review

enriqueartal committed:
use libbraiding algorithm for left_normal_form

enriqueartal committed:
indent

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1447,6 +1448,52 @@ def annular_khovanov_homology(self, qagrad=None, ring=IntegerRing()):
+        
+        - `algorithm` : string (default: ``artin``). For `artin` the general
+          method for Artin group is used. For `libbraiding`, the algorithm
+          uses the package `libbraiding`.
```
```suggestion
        - ``algorithm`` -- string (default: ``'artin'``); must be one of the following:

          * ``'artin'`` -- the general method for Artin groups is used
          * ``'libbraiding'`` -- the algorithm uses the ``libbraiding`` package
```
Why also a default of `artin` instead of `libbraiding` since it is faster? I was only suggesting keeping both available, but you can certainly change the default.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1447,6 +1448,52 @@ def annular_khovanov_homology(self, qagrad=None, ring=IntegerRing()):
+            return tuple([P.delta() ** lnfp[0]] +
+                        [P._standard_lift(w) for w in lnfp[1:]])
+
+
```
```suggestion
            return tuple([P.delta() ** lnfp[0]] +
                         [P._standard_lift(w) for w in lnfp[1:]])

```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1657,6 +1739,91 @@ def is_conjugated(self, other):
+        
+        A pure braid `d` such that `other` equals `d^-1 * self * d`.
+         
+
```
Remove one of the blanklines here.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -89,7 +90,7 @@
 
 lazy_import('sage.libs.braiding',
-            ['rightnormalform', 'centralizer', 'supersummitset', 'greatestcommondivisor',
+            ['leftnormalform','rightnormalform', 'centralizer', 'supersummitset', 'greatestcommondivisor',
```
```suggestion
            ['leftnormalform', 'rightnormalform', 'centralizer', 'supersummitset', 'greatestcommondivisor',
```

tscrim reviewed (changes requested):
Thank you. It is looking good overall; just a few more details.

So it is fine to change more files. It is up to you how large you want to make the scope of the changes. Adding such an option to `permutation()` and `coxeter_group_element()` would definitely be an improvement, but again, not necessarily to do here (but you can if you want to).

enriqueartal committed:
simplify and cached left_normal_form

enriqueartal committed:
permutation method output in SymmetricGroup

enriqueartal committed:
change == by is

enriqueartal committed:
change the default of permutation method

enriqueartal committed:
style typo

enriqueartal committed:
more style typos

enriqueartal committed:
typo finally found

enriqueartal committed:
documentation

enriqueartal committed:
typo in pure_conjugating_braid

enriqueartal committed:
typo in doc

miguelmarco commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1680,55 @@ def conjugating_braid(self, other):
         else:
-            return self.parent()._element_from_libbraiding(l)
+            B = self.parent()
+            n = B.strands()
```
what is the reason for getting the number of strands?

enriqueartal replied:
To define the symmetric group

tscrim replied:
You no longer define it here.

enriqueartal committed:
use tscrim code

enriqueartal committed:
Merge branch 'sagemath:develop' into braids

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -427,24 +428,34 @@ def alexander_polynomial(self, var='t', normalized=True):
         return p
 
-    def permutation(self):
+    def permutation(self, W = None):
```
```suggestion
    def permutation(self, W=None):
```
PEP8

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -427,24 +428,34 @@ def alexander_polynomial(self, var='t', normalized=True):
+
+        - ``W`` -- (default : ``None``); the ``SymmetricGroup(n)``
+          where ``n`` is the number of strands of ``self`` is
+          an alternative option.
```
```suggestion
        - ``W`` -- (optional) the permutation group to project
          ``self`` to; the default is ``self.parent().coxeter_group()``
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -427,24 +428,34 @@ def alexander_polynomial(self, var='t', normalized=True):
 
-        A permutation.
+        A standard permutation if ``W`` is ``None``, an element of ``W`` if it
+        is ``SymmetricGroup(n)``.
```
```suggestion
        The image of ``self`` under the natural projection map to ``W``.
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -427,24 +428,34 @@ def alexander_polynomial(self, var='t', normalized=True):
-            sage: b.permutation()
+            sage: b.permutation(W=S)
+            (1,4,2)
+            sage: c = Permutation(b.permutation(W=S)); c
```
```suggestion
            sage: c = b.permutation(W=Permutations(4)); c
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1447,6 +1458,51 @@ def annular_khovanov_homology(self, qagrad=None, ring=IntegerRing()):
+
+        - ``algorithm`` : string (default: ``libbraiding``). For ``artin`` the general
+          method for Artin group is used. For ``libbraiding``, the algorithm
+          uses the package ``libbraiding``.
```
```suggestion
        - ``algorithm`` -- string (default: ``'artin'``); must be one of the following:

          * ``'artin'`` -- the general method for Artin groups is used
          * ``'libbraiding'`` -- the algorithm from the ``libbraiding`` package
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1447,6 +1458,51 @@ def annular_khovanov_homology(self, qagrad=None, ring=IntegerRing()):
+            return tuple([B.delta()**l[0][0]] + [B(b) for b in l[1:]] )
+        elif algorithm == 'artin':
+            return FiniteTypeArtinGroupElement.left_normal_form.f(self)
+
```
Raise an error if we get here saying invalid algorithm.

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1471,6 +1527,10 @@ def _left_normal_form_coxeter(self):
 
+        .. NOTE::
+
+            For long braids this method is slower than leftnormalform from `libbraiding`.
```
```suggestion
            For long braids this method is slower than ``algorithm='libbraiding'``.
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1678,55 @@ def conjugating_braid(self, other):
         INPUT:
 
-        - ``other`` -- the other braid to look for conjugating braid
+        - ``other`` -- a braid in the same braid group as ``self``.
```
```suggestion
        - ``other`` -- a braid in the same braid group as ``self``
```

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1658,6 +1750,91 @@ def is_conjugated(self, other):
+            sage: d1 * a / d1 == c
+            False
+            sage: d2 = a.pure_conjugating_braid(c)
+            sage: print(len(d2.Tietze()))
```
You don't need all these `print`s here.

enriqueartal replied:
Done for last suggestions, I will upload the changes soon

tscrim reviewed (changes requested):
Still a little bit more work to do, but most of it is trivial at this point.

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Update src/sage/groups/braid.py

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
tests

enriqueartal committed:
Merge branch 'braids' of github.com:enriqueartal/sage into braids

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1501,7 +1500,8 @@ def left_normal_form(self, algorithm='libbraiding'):
             return FiniteTypeArtinGroupElement.left_normal_form.f(self)
-
+        else:
+            raise ValueError("invalid algorithm")
```
```suggestion
        raise ValueError("invalid algorithm")
```
This is just my personal coding style, but I think it makes it a bit more apparent (and catches more bugs) if it is just a fall through. That way the method will not secretly return `None` and each branch of the `if` statement returns something.

enriqueartal committed:
Update src/sage/groups/braid.py

Avoid an extra `else`

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

enriqueartal committed:
Added more examples for left_normal_form, conjugating_braid, and pure_conjugating_braid,
drop useless print commands, and put mathematical descriptions of conjugating_braid and
pure_conjugating_braid in separate paragraphs

tscrim reviewed (approved):
Thank you. LGTM.

enriqueartal committed:
There were typos in the docs of reduced_burau_matrix, jones_polynomial, TL_representation and links_gould_polynomial

enriqueartal committed:
lint problems

enriqueartal committed:
Merge branch 'sagemath:develop' into braids

tscrim commented on src/sage/groups/braid.py:
```diff
@@ -1614,27 +1692,63 @@ def conjugating_braid(self, other):
+        A conjugating braid.
+
+        More precisely, if the output is `d`, `o` equals ``other``, and `s` equals ``self``
+        then `o = d^{-1} \cdot s \cdot d`.
```
One last trivial thing: This needs a blankline after it.

enriqueartal replied:
Done, and also style spaces all around

enriqueartal committed:
Added blank line and plenty of style spaces

enriqueartal committed:
Some lint typos

tscrim reviewed (changes requested):
A lot of those were not PEP8 violations (higher precedence operators do not need the same spacing, which include `[]` and `()`), and some of them actually makes the code harder to read. However, the rule was not applied consistently. I would actually just revert those additional PEP8 spacing changes as it makes the patch larger for minimal (at best) gain. Better separation of ticket purpose.

enriqueartal committed:
Revert style changes, except blank line

enriqueartal committed:
Merge branch 'braids' of github.com:enriqueartal/sage into braids

enriqueartal committed:
Keep only required blank line

tscrim reviewed (approved):
Thank you. I think this is ready to go in!
